### ISSUE TITLE
Torch three dimensional matmul operation is ~875% slower than conventional two dimensional matrix multiplication

### ISSUE THREAD

**Issue by blint071** [pytorch/pytorch#58760]
The thing is that the issue comes when doing the matmul operation with two three dimensional arrays as arguments. 
Sorry if the bug report was a little misleading, when the dimension 1 of the first argument equals to one, it can also be considered as a vector by matrix multiplication.

```python
a=torch.rand(size=(100000,1,100), dtype=torch.float16, device='cuda')
b=torch.rand(size=(100000,100,100), dtype=torch.float16, device='cuda')
 
torch.cuda.synchronize()
 
start = torch.cuda.Event(enable_timing=True)
end = torch.cuda.Event(enable_timing=True)
 
start.record()
z = torch.matmul(a,b)
end.record()
 

torch.cuda.synchronize()
 
print(start.elapsed_time(end))
 
 
a=torch.rand(size=(200000,100), dtype=torch.float16, device='cuda')
b=torch.rand(size=(100,100), dtype=torch.float16, device='cuda')
 
torch.cuda.synchronize()
 
start = torch.cuda.Event(enable_timing=True)
end = torch.cuda.Event(enable_timing=True)
 
start.record()
z = torch.matmul(a,b)
end.record()
 

torch.cuda.synchronize()
 
print(start.elapsed_time(end))
```

My results:
```
Three dimensional matmul op: 12.984319686889648
Two dimensional matmul op: 2.070528030395508
 ```

Correct me if I am wrong, but this should not be the expected performance, the number of elements that the gpu has to access remains the same in both operations. What is more, the number of multiplications/summations is even higher when doing the two dimensional matmul operation and the performance difference is pretty much noticeable.
 
Going more in depth into the problem.
What might also be difficult to understand is that, when increasing the batch dim 1 of the first argument in the three dimensional matmul op, there is almost no performance impact, even though the number of operations is considerably higher.
However, when increasing batch dim 0 or 2 in the same argument, the performance impact scales linearly. 
 
This makes me believe that the problem may rely on the lack of parallelization of dim 0 when doing this kind of high dimensional matmul operations, but I don't really know.
 
I think it would be better to completely edit the issue, I also think it was difficult to understand.

## Environment
```
Collecting environment information...
PyTorch version: 1.8.1+cu101
Is debug build: False
CUDA used to build PyTorch: 10.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.12.0
Libc version: glibc-2.26

Python version: 3.7 (64-bit runtime)
Python platform: Linux-5.4.109+-x86_64-with-Ubuntu-18.04-bionic
Is CUDA available: True
CUDA runtime version: 11.0.221
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 460.32.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.4
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] torch==1.8.1+cu101
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.9.1
[pip3] torchvision==0.9.1+cu101
[conda] Could not collect
```

cc @jianyuh @nikitaved @pearu @mruberry @heitorschueroff @walterddr @IvanYashchuk @xwang233 @Lezcano @rgommers @VitalyFedyunin @ngimel

**Comment by rgommers** [comment]
@blint071 your question is hard to understand, can you please narrow it down to one single operation (`matmul`, not vector-matrix multiply), and compare a single batched vs. non-batched call?

> As I said the number of elements and summations/multiplications is higher in op4 than in op1, so at least similar performance should be expected in op1.

It would be nice if this expectation was correct, but life isn't so simple unfortunately. A `matmul` call can be two orders of magnitude faster than a naive "do multiplies and adds in C/C++".

**Comment by blint071** [comment]
Updated the issue, now the problem is more clear.

**Comment by rgommers** [comment]
Thanks. I simplified it a bit more and added output shapes:
```python

In [10]: a = torch.rand(size=(100000,1,100), dtype=torch.float16)

In [11]: b = torch.rand(size=(100000,100,100), dtype=torch.float16)

In [12]: b.shape
Out[12]: torch.Size([100000, 100, 100])

In [13]: torch.matmul(a, b).shape
Out[13]: torch.Size([100000, 1, 100])

In [14]: c = torch.rand(size=(200000,100), dtype=torch.float16)

In [15]: d = torch.rand(size=(100,100), dtype=torch.float16)

In [16]: torch.matmul(c, d).shape
Out[16]: torch.Size([200000, 100])
```

You are comparing a case with `100_000` batched matmuls of shapes `(1, 100) x (100, 100)` with a case of a single matmul of shapes `(200_000, 100) x (100, 100)`. I don't know why there's a factor 2 difference in `100_000` vs. `200_000` in your example, but either way I don't have any expectation that the timings for these two cases should be exactly equal.

**Comment by blint071** [comment]
The reason for the factor of two is to demonstrate that the bottleneck is not in the GPU memory bandwidth, as someone could argue that the performance difference could be caused by the GPU having to access a higher number of elements. 
 
But again, I don't understand why so much attention to that. The second matrix operation literally has access to twice the number of elements than the first one and also perform twice the number of operations, and it still performs 550% better.
The difference is still there when doing the benchmark with the same number of operations.
```
a=torch.rand(size=(100000,1,100), dtype=torch.float16, device='cuda')
b=torch.rand(size=(100000,100,100), dtype=torch.float16, device='cuda')
 
a=torch.rand(size=(100000,100), dtype=torch.float16, device='cuda')
b=torch.rand(size=(100,100), dtype=torch.float16, device='cuda')
```
My results:
```
13.146112442016602
1.132032036781311
```

**Comment by blint071** [comment]
Update: I tested different CUDA kernels which NVIDIA says are  expecialliced  for batched matrix multiplication, “many (typically small) matrix-matrix multiplies at once”. 
I don't want to go into specific details but it seems that this issue doesn´t have an easy fix.
Both kernels, cublasHgemmBatched and cublasHgemmStridedBatched failed to get better performance than pytorch with the same benchmarking parameters, with the first one being two times slower and the second one being slightly slower than the official torch matmul.
This could mean that both users made a bad usage of the cublas API, that I failed to port the cuda code or that the “expecialliced  for batched matrix multiplication kernels” kernels were not that expecialliced, as the performance was still 8 times worse than 2D matrix multiplication(or non-batched).

Props to qtqt(https://stackoverflow.com/questions/48519861/strange-cublas-gemm-batched-performance) and Robert Crovella(https://stackoverflow.com/questions/23743384/how-performing-multiple-matrix-multiplications-in-cuda).

I had to modify both of them because one of them simply didn't work and the other one only worked with a batch dim of 2. I uploaded both implementations(crap but functional code) in my github https://github.com/blint071, you need to time and call sync from outside and subtract the time used to generate and allocate tensor arrays.

There have been some posts complaining about this issue during these years but none of them actually led to any solution. I could copy paste this to the tensorflow or nvidia forums, but it would probably also lead to dead discussions.

As you could guess, the only way to fix this mess would be to make a raw kernel from the start to end, as none of the cublas official kernels seems to work right, this is something that only someone with a high knowledge of CUDA could do.

This operation is not strange, I can show some examples of how this is being used in transformers with multi head attention.

Considering that matmul is one of the most important operations in ML, and that the performance gets reduced in that way just by using 3 dimensional matrices instead of 2, I would consider it as a high priority issue.

**Comment by ngimel** [comment]
On my V100 system the operation takes ~3.7ms according to the benchmark above. One needs to read matrix b, which is (100000*100*100*2bytes/element) bytes. Reading of a and writing result are much faster so let's ignore them. This gives a bandwidth of 100000*100*100*2*1e-9/3.7e-3 = 540 GB/s. V100 maximum memory bandwidth is 900 GB/s, so a perfect implementation, if it can be achieved, would speed up this operation by 40% at best. There's no way this matrix multiplication would be as fast as 2d one, because that one requires reading much less data, and realistically, current implementation can't even be sped up that much.

**Comment by blint071** [comment]
Yes you are right, I miscalculated some things. Sorry
